src/psykal/psy.f90

```diff
@@ -34,150 +26,171 @@ subroutine tracer_advection(zind,tsn,ztfreez,rnfmsk,rnfmsk_z,upsmsk, &
     use zslpx_update_kern_mod, only : zslpx_update_kern
     use zwx2_kern_mod, only : zwx2_kern
     use mydomain_kern_mod, only : mydomain_kern
+    implicit none
     !
     ! RF all intent out's here are actually local to this call
-    real*8, intent(out) :: zind(jpi,jpj,jpk), zwx(jpi,jpj,jpk), &
-         zwy(jpi,jpj,jpk), zslpx(jpi,jpj,jpk), zslpy(jpi,jpj,jpk)
-    real*8, intent(in)  :: tsn(jpi,jpj,jpk), ztfreez(jpi,jpj), &
-         rnfmsk(jpi,jpj), rnfmsk_z(jpk), upsmsk(jpi,jpj), tmask(jpi,jpj,jpk), &
-         umask(jpi,jpj,jpk), vmask(jpi,jpj,jpk), pun(jpi,jpj,jpk), &
-         pvn(jpi,jpj,jpk), pwn(jpi,jpj,jpk)
-    real*8, intent(inout) :: mydomain(jpi,jpj,jpk)
+    type(r3d_field), intent(inout) :: zind, zwx, zwy, zslpx, zslpy
+    type(r3d_field), intent(in)    :: tsn, pun, pvn, pwn
+    type(r2d_field), intent(in)    :: rnfmsk, upsmsk, ztfreez
+    type(r1d_field), intent(in)    :: rnfmsk_z
+    type(r3d_field), intent(inout) :: mydomain
     ! local variables
     integer :: ji,jj,jk
-    !
+    type(grid_type), pointer :: grid
+    !
+    ! Get a pointer to the grid from a read-only argument. We then use
+    ! this to access grid-properties (such as masks) required by the
+    ! kernels
+    grid => tsn%grid
+
+    jpi = grid%simulation_domain%xstop
+    jpj = grid%simulation_domain%ystop
+    jpk = grid%nlevels
```

Comment:
Yes and no. NEMO does have a fixed/constant `jpk` but, depending on the configuration, it can use a 2D array that holds the index of the last wet level for any `i,j` location. Since that isn't used in the tracer kernel I have again left it as future work :-)